### PR TITLE
Implicit onDrain() when using db.query()

### PR DESCRIPTION
**Pull request by staltz**
Implementation comments are inline in the code

### Performance before, no EBT index

```
➜  ssb-db2 git:(master) rm -rf ~/.ssb/db2/indexes

➜  ssb-db2 git:(master) node gitignored/example.js
There are 15 messages
duration: 22285ms

➜  ssb-db2 git:(master) node gitignored/example.js
There are 15 messages
duration: 419ms

➜  ssb-db2 git:(master) node gitignored/example.js
There are 15 messages
duration: 401ms

➜  ssb-db2 git:(master) node gitignored/example.js
There are 15 messages
duration: 402ms

➜  ssb-db2 git:(master) du -h ~/.ssb/db2/indexes
7,6M	/home/staltz/.ssb/db2/indexes/base
29M	/home/staltz/.ssb/db2/indexes
```

### Performance after, no EBT index

```
➜  ssb-db2 git:(ondrain) rm -rf ~/.ssb/db2/indexes

➜  ssb-db2 git:(ondrain) node gitignored/example.js
There are 15 messages
duration: 20885ms

➜  ssb-db2 git:(ondrain) node gitignored/example.js
There are 15 messages
duration: 464ms

➜  ssb-db2 git:(ondrain) node gitignored/example.js
There are 15 messages
duration: 407ms

➜  ssb-db2 git:(ondrain) node gitignored/example.js
There are 15 messages
duration: 406ms

➜  ssb-db2 git:(ondrain) du -h ~/.ssb/db2/indexes
7,6M /home/staltz/.ssb/db2/indexes/base
29M  /home/staltz/.ssb/db2/indexes
```

### CODE REVIEW

File: compat/history-stream.js [hunk 1]
```diff
@@ -50,17 +50,19 @@ exports.init = function (sbot, config) {
 
     return pull(
       pullCont(function (cb) {
-        if (!ref.isFeed(opts.id)) return cb(opts.id + ' is not a feed')
+        sbot.db.log.onDrain(() => {
```
Comment (staltz) [inline compat/history-stream.js:53]: onDrain was needed here because we're not using `query` from `db.js` which has the implicit onDrain. Notice I had to export the log in `sbot.db.log`

File: db.js [hunk 2]
```diff
@@ -70,16 +69,14 @@ exports.init = function (sbot, dir, config) {
...
-      query(
-        and(key(id)),
-        toCallback((err, results) => {
-          if (err) return cb(err)
-          else if (results.length) return cb(null, results[0].value)
-          else return cb()
-        })
-      )
-    })
+    query(
```
Comment (staltz) [inline db.js:72]: onIndexesStateLoaded was not needed because that's about base index, ebt index, private index, etc, and this query only uses JITDB things. It will use an implicit onDrain because it's `query`

File: db.js [hunk 3]
```diff
@@ -351,29 +348,26 @@ exports.init = function (sbot, dir, config) {
 
   // override query() from jitdb to implicitly call fromDB()
   function query(first, ...rest) {
+    const waitForDrain = and(deferred((meta, cb) => log.onDrain(cb)))
```
Comment (staltz) [inline db.js:351]: Here is the implicit onDrain

File: operators/full-mentions.js [hunk 4]
```diff
@@ -2,6 +2,10 @@ const { deferred } = require('jitdb/operators')
 
 module.exports = function mentions(key) {
   return deferred((meta, cb) => {
-    meta.db2.getIndexes().fullMentions.getMessagesByMention(key, meta.live, cb)
+    meta.db2.onDrain('fullMentions', () => {
+      meta.db2
+        .getIndexes()
+        .fullMentions.getMessagesByMention(key, meta.live, cb)
+    })
```
Comment (staltz) [inline operators/full-mentions.js:9]: This one was actually a great idea! The custom operator can do the onDrain for the specific index that it is meant to, in this case `full-mentions`.

File: test/createHistoryStream.js [hunk 5]
```diff
@@ -126,7 +122,7 @@ test('limit', (t) => {
         pull(
           sbot.createHistoryStream({ id: keys.id }),
           pull.collect((err, results) => {
-            t.equal(results.length, 2)
+            t.equal(results.length, 3)
```
Comment (staltz) [inline test/createHistoryStream.js:125]: Notice that this test is fundamentally testing something different now. I think the previous was actually testing the wrong thing, and it kept on passing because this test **didn't** have an onDrain. I took a careful look at what the published messages were, and I think there are 3.

File: test/full-mentions.js [hunk 6]
```diff
@@ -48,30 +48,25 @@ test('getMessagesByMention', (t) => {
       db.publish(mentionMsg, (err) => {
         t.error(err, 'no err')
 
-        db.onDrain('fullMentions', () => {
-          const status = db.getStatus()
-          t.equal(status.indexes['fullMentions'], 780, 'index in sync')
```
Comment (staltz) [inline test/full-mentions.js:53]: This isn't using `getStatus` anymore, because that would have to happen after the `onDrain('fullMentions'`, but the onDrain is done implicitly inside the custom operator `mentions()`, so I didn't know where to put this one...

File: test/migration.js [hunk 7]
```diff
@@ -61,24 +61,27 @@ test('migrate moves msgs from old log to new log', (t) => {
     fromEvent('ssb:db2:migrate:progress', sbot),
     pull.filter((x) => x === 1),
     pull.take(1),
-    pull.drain(() => {
-      // we need to make sure async-flumelog has written the data
-      sbot.db.onDrain(() => {
+    // we need to make sure async-flumelog has written the data
+    pull.asyncMap((_, continueCb) =>
+      setTimeout(() => {
```
Comment (staltz) [inline test/migration.js:66]: This setTimeout is just waiting for the fs to be written, seems unrelated to onDrain

File: db.js (hunk 3 above)
Comment (arj03) [reply db.js:351]: This is pretty clever

File: test/createHistoryStream.js (hunk 5 above)
Comment (arj03) [reply test/createHistoryStream.js:125]: Yes

File: test/full-mentions.js (hunk 6 above)
Comment (arj03) [reply test/full-mentions.js:53]: should be fine, we really just care about the results

File: test/migration.js (hunk 7 above)
Comment (arj03) [reply test/migration.js:66]: can't you do a sbot.db.log.onDrain? If you just care about when stuff has been written, that seems like a safer bet

File: test/operators.js [hunk 8]
```diff
@@ -296,31 +282,31 @@ test('live votesFor', (t) => {
...
+        db.query(
+          and(votesFor(postMsg.key)),
+          live(),
+          toPullStream(),
+          pull.drain(
+            (result) => {
+              if (i++ == 0) {
+                t.equal(result.key, v1.key)
+
+                setTimeout(() => {
```
Comment (arj03) [inline test/operators.js:296]: why the settimeout here?

**Comment by arj03** [comment]
Added some minor comments, no rush with this, just wanted to get the notes in so I don't have to think about this :)

File: test/operators.js (hunk 8 above)
Comment (staltz) [reply test/operators.js:296]: Oh, I think that's an artifact of debugging left around. I'll try to remove it

File: test/migration.js (hunk 7 above)
Comment (staltz) [reply test/migration.js]: So, I wrote this setTimeout hack **before** I decided to export `log` at `sbot.db.log`. Now that we have it, it makes more sense to use `sbot.db.log.onDrain` indeed.

**Comment by staltz** [comment]
Was easy to fix! https://github.com/ssb-ngi-pointer/ssb-db2/pull/79/commits/6a5f7d13ebf96324219f3c4d90645905d4aa38d0
